src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,309 @@
+                    return $this->offsetGet($offset);
+                },
+            ),
+        );
+    }
+
+    public function jsonSerialize(): array
+    {
+        return iterator_to_array($this->getIterator());
+    }
+
+    /** @param mixed $offset */
+    public function offsetExists($offset): bool
+    {
+        if (! is_numeric($offset)) {
+            return false;
+        }
+
+        $offset = (int) $offset;
+
+        return $this->exists[$offset] ??= $this->bson->has($offset);
+    }
+
+    /**
+     * @param mixed $offset
+     * @return TValue
+     */
+    #[ReturnTypeWillChange]
+    public function offsetGet($offset)
+    {
+        if (! is_numeric($offset)) {
+            trigger_error(sprintf('Undefined offset: %s', $offset), E_USER_WARNING);
+
+            return null;
+        }
+
+        $offset = (int) $offset;
+        $this->readFromBson($offset);
+
+        if (isset($this->unset[$offset]) || ! $this->exists[$offset]) {
+            trigger_error(sprintf('Undefined offset: %d', $offset), E_USER_WARNING);
+
+            return null;
+        }
+
+        return array_key_exists($offset, $this->set) ? $this->set[$offset] : $this->read[$offset];
+    }
+
+    /**
+     * @param mixed  $offset
+     * @param TValue $value
+     */
+    public function offsetSet($offset, $value): void
+    {
+        if ($offset === null) {
+            $this->readEntirePackedArray();
+
+            $existingItems = [
+                ...array_keys($this->read),
+                ...array_keys($this->set),
+            ];
+
+            $offset = $existingItems === [] ? 0 : max($existingItems) + 1;
+        } elseif (! is_numeric($offset)) {
+            trigger_error(sprintf('Unsupported offset: %s', $offset), E_USER_WARNING);
+
+            return;
+        } else {
+            $offset = (int) $offset;
+        }
+
+        $this->set[$offset] = $value;
+        unset($this->unset[$offset]);
+        $this->exists[$offset] = true;
+    }
+
+    /** @param mixed $offset */
+    public function offsetUnset($offset): void
+    {
+        if (! is_numeric($offset)) {
+            trigger_error(sprintf('Undefined offset: %s', $offset), E_USER_WARNING);
+
+            return;
+        }
+
+        $offset = (int) $offset;
+        $this->unset[$offset] = true;
+        $this->exists[$offset] = false;
+        unset($this->set[$offset]);
+    }
+
+    private function readEntirePackedArray(): void
+    {
+        if ($this->entirePackedArrayRead) {
+            return;
+        }
+
+        foreach ($this->bson as $offset => $value) {
+            $this->read[$offset] = $value;
+
+            if (! isset($this->exists[$offset])) {
+                $this->exists[$offset] = true;
+            }
+        }
+
+        $this->entirePackedArrayRead = true;
+    }
+
+    private function readFromBson(int $offset): void
+    {
+        if (array_key_exists($offset, $this->read)) {
```

Comment:
I'll keep that in mind to test in the performance PR - while it may seem like a micro-optimisation this is very much in the hot path, so it may be worth it.

Reply:
Will defer to you to create a ticket for this (dependent on benchmarks being implemented). Feel free to resolve.

Reply:
Created PHPLIB-1212.